core/src/main/java/org/elasticsearch/action/admin/indices/shrink/TransportShrinkAction.java

```diff
@@ -0,0 +1,210 @@
+            throw new IllegalArgumentException("mappings are not allowed when shrinking indices" +
+                ", all mappings are copied from the source index");
+        }
+        final Settings tragetIndexSettings = Settings.builder().put(targetIndex.settings())
+            .normalizePrefix(IndexMetaData.INDEX_SETTING_PREFIX).build();
+        if (IndexMetaData.INDEX_NUMBER_OF_SHARDS_SETTING.exists(tragetIndexSettings)
+            && IndexMetaData.INDEX_NUMBER_OF_SHARDS_SETTING.get(tragetIndexSettings) > 1) {
+            throw new IllegalArgumentException("can not shrink index into more than one shard");
+        }
+        // now check that index is all on one node
+        final IndexRoutingTable table = state.routingTable().index(sourceIndex);
+        Map<String, AtomicInteger> nodesToNumRouting = new HashMap<>();
+        int numShards = metaData.getNumberOfShards();
+        for (ShardRouting routing : table.shardsWithState(ShardRoutingState.STARTED)) {
+            nodesToNumRouting.computeIfAbsent(routing.currentNodeId(), (s) -> new AtomicInteger(0)).incrementAndGet();
+        }
+        List<String> nodesToAllocateOn = new ArrayList<>();
+        for (Map.Entry<String, AtomicInteger> entries : nodesToNumRouting.entrySet()) {
+            int numAllocations = entries.getValue().get();
+            assert numAllocations <= numShards : "wait what? " + numAllocations + " is > than num shards " + numShards;
+            if (numAllocations == numShards) {
+                nodesToAllocateOn.add(entries.getKey());
+            }
+        }
+        if (nodesToAllocateOn.isEmpty()) {
+            throw new IllegalStateException("index " + sourceIndex +
+                " must have all shards allocated on the same node to shrink index");
+        }
+        targetIndex.cause("shrink_index");
+        Predicate<String> analysisSimilarityPredicate = (s) -> s.startsWith("index.similarity.") || s.startsWith("index.analysis.");
+        targetIndex.settings(Settings.builder()
+            .put(tragetIndexSettings)
+            // we can only shrink to 1 index so far!
+            .put("index.number_of_shards", 1)
+            // we set default to 0 only if there is nothing explicitly set
+            .put("index.number_of_replicas", IndexMetaData.INDEX_NUMBER_OF_REPLICAS_SETTING.exists(tragetIndexSettings) ?
+                IndexMetaData.INDEX_NUMBER_OF_REPLICAS_SETTING.get(tragetIndexSettings) : 0)
+            // we use "i.r.a.include" rather than "i.r.a.require" since it's allows one of the nodes holding an instance of all shards.
+            .put("index.routing.allocation.include._id", Strings.arrayToCommaDelimitedString(nodesToAllocateOn.toArray()))
+            // we only try once and then give up with a shrink index
+            .put("index.allocation.max_retries", 1)
+            // now copy all similarity / analysis settings - this overrides all settings from the user unless they wanna add extra settings
```

Comment:
I wonder if we should be stricter and require no analysisSimilarity settings are on the target node - o.w. who knows what kind of mixture we'll get? Thinking more about it - should we just white list the settings that are OK to change? like compression and grow the list as we go?